File: packages/vue-im-ui/src/shared/contracts/message-lifecycle.ts
import type { TransferState } from "./transfer";

export type MessageSendState = "draft" | "sending" | "sent" | "failed";
export type MessageDeliveryState = "serverAccepted" | "delivered" | "partiallyDelivered";
export type MessageReadState = "unread" | "partiallyRead" | "read";
export type MessageMutationState = "normal" | "edited" | "recalled" | "deleted";
export type MessageEphemeralState = "none" | "readOnce" | "burnAfterRead" | "expired";

/** UI-facing lifecycle projection. Hosts map authoritative state into this model. */
export interface MessageLifecycle {
  transfer: TransferState;
  send: MessageSendState;
  delivery: MessageDeliveryState;
  read: MessageReadState;
  mutation: MessageMutationState;
  ephemeral: MessageEphemeralState;
}

/** Visual projection only. The orthogonal lifecycle remains the source model. */
export type MessageStatusState =
  | "pending"
  | "sending"
  | "sent"
  | "delivered"
  | "read"
  | "failed"
  | "retrying";

export const defaultMessageLifecycle: MessageLifecycle = {
  transfer: "idle",
  send: "sent",
  delivery: "serverAccepted",
  read: "unread",
  mutation: "normal",
  ephemeral: "none",
};

/** Maps lifecycle dimensions into the final receipt glyph without collapsing the model. */
export function lifecycleToMessageStatus(lifecycle: MessageLifecycle): MessageStatusState {
  if (lifecycle.send === "failed" || lifecycle.transfer === "failed") return "failed";
  if (lifecycle.send === "draft" || lifecycle.transfer === "queued") return "pending";
  if (
    lifecycle.send === "sending" ||
    lifecycle.transfer === "transferring" ||
    lifecycle.transfer === "paused"
  ) return "sending";
  if (lifecycle.read === "read") return "read";
  if (
    lifecycle.delivery === "delivered" ||
    lifecycle.delivery === "partiallyDelivered"
  ) return "delivered";
  return "sent";
}

File: packages/flutter-im-ui/lib/src/models/message_lifecycle.dart
import '../components/flare_message_status.dart';
import '../components/flare_transfer_progress.dart';

enum FlareMessageSendState { draft, sending, sent, failed }

enum FlareMessageLifecycleDeliveryState {
  serverAccepted,
  delivered,
  partiallyDelivered,
}

enum FlareMessageReadState { unread, partiallyRead, read }

enum FlareMessageMutationState { normal, edited, recalled, deleted }

enum FlareMessageEphemeralState { none, readOnce, burnAfterRead, expired }

/// UI-facing lifecycle projection mapped from authoritative host state.
class FlareMessageLifecycle {
  const FlareMessageLifecycle({
    this.transfer = FlareTransferState.idle,
    this.send = FlareMessageSendState.sent,
    this.delivery = FlareMessageLifecycleDeliveryState.serverAccepted,
    this.read = FlareMessageReadState.unread,
    this.mutation = FlareMessageMutationState.normal,
    this.ephemeral = FlareMessageEphemeralState.none,
  });

  final FlareTransferState transfer;
  final FlareMessageSendState send;
  final FlareMessageLifecycleDeliveryState delivery;
  final FlareMessageReadState read;
  final FlareMessageMutationState mutation;
  final FlareMessageEphemeralState ephemeral;

  FlareMessageDeliveryStatus get visualStatus {
    if (send == FlareMessageSendState.failed ||
        transfer == FlareTransferState.failed) {
      return FlareMessageDeliveryStatus.failed;
    }
    if (send == FlareMessageSendState.draft ||
        transfer == FlareTransferState.queued) {
      return FlareMessageDeliveryStatus.pending;
    }
    if (send == FlareMessageSendState.sending ||
        transfer == FlareTransferState.transferring ||
        transfer == FlareTransferState.paused) {
      return FlareMessageDeliveryStatus.sending;
    }
    if (read == FlareMessageReadState.read)
      return FlareMessageDeliveryStatus.read;
    if (delivery == FlareMessageLifecycleDeliveryState.delivered ||
        delivery == FlareMessageLifecycleDeliveryState.partiallyDelivered) {
      return FlareMessageDeliveryStatus.delivered;
    }
    return FlareMessageDeliveryStatus.sent;
  }
}
